qiime2/metadata/metadata.py

```diff
@@ -85,7 +87,123 @@ def artifacts(self):
             Source artifacts of the metadata.
 
         """
-        return tuple(self._artifacts)
+        artifacts = []
+        for source in self._sources:
+            if isinstance(source, qiime2.Artifact):
+                artifacts.append(source)
+        return tuple(artifacts)
+
+    # This is to be gotten rid of
+    @property
+    def non_artifacts(self):
+        """Artifacts that are the source of the metadata.
+
+        This property is read-only.
+
+        Returns
+        -------
+        tuple of qiime2.Artifact
+            Source artifacts of the metadata.
+
+        """
+        artifacts = []
+        for source in self._sources:
+            if isinstance(source, qiime2.Metadata):
+                artifacts.append(source)
+        return tuple(artifacts)
+
+    @property
+    def dataframes(self):
+        """DataFrames that are the source of the metadata.
+
+        This property is read-only.
+
+        Returns
+        -------
+        tuple of qiime2.Metadata
+            Source metadata from DataFrames of the metadata.
+
+        """
+        dataframes = []
+        for source in self._sources:
+            if source._source_type == 'DataFrame':
+                dataframes.append(source)
+        return tuple(dataframes)
+
+    @property
+    def tsvs(self):
+        """tsvs that are the source of the metadata.
+
+        This property is read-only.
+
+        Returns
+        -------
+        tuple of qiime2.Metadata
+            Source metadata from tsvs of the metadata.
+
+        """
+        tsvs = []
+        for source in self._sources:
+            if source._source_type == 'tsv':
+                tsvs.append(source)
+        return tuple(tsvs)
+
+    @property
+    def merges(self):
+        """Metadatas resulting from previous merges that are the source of the
+        metadata.
+
+        This property is read-only.
+
+        Returns
+        -------
+        tuple of qiime2.Metadata
+            Source metadata from previous merges of the metadata.
+
+        """
+        merges = []
+        for source in self._sources:
+            if source._source_type == 'Merge':
+                merges.append(source)
+        return tuple(merges)
+
+    @property
+    def sources(self):
+        """All sources of the metadata.
+
+        This property is read-only.
+
+        Returns
+        -------
+        tuple of qiime2.Metadata and qiime2
+            Source metadata of this metadata.
+
+        """
+        return tuple(self._sources)
+
+    @property
+    def source_type(self):
```

Comment:
Do we need this? How does this work when the current MD instance is the result of a merging operation?

Reply:
The source type is "merge" without this and with the single pooled `_source` attribute how would we know specifically what the source of a given object was?

Reply:
I think I suggested _replacing_ `_source` with these new partitioned source properties, so that you can still keep track of where things came from without quite so much effort. I will write up my thoughts in more detail later today.